openedx/edx-enterprise
# feat: create cea for invite only courses before checkout

**NOTE** The PR has been moved to draft to rework the changes.
### Description

This allows enterprise learners whose enterprise catalog contains private "invitation only" courses to enroll without having to be explicitly invited. 

This adds a new flag to the `EnterpriseCustomer` models called `allow_enrollment_in_invite_only_courses`. Once this is enabled, the enrollment POST request will check if the course is "invite-only" and create a `CourseEnrollmentAllowed` object in the platform if required.

### Testing instructions

The following instruction assume you are using the devstack setup for testing.

1. In a working edx-enterprise setup, add a course which is "invite-only" to the catalog, or change an existing course in the catalog to "Invitation Only". This can be done by going to the "Advanced Settings" in Studio and setting the "Invitation Only" field to `true`.
2. Get the enrollment_url of the course - Usually something like this `http://localhost:18000/enterprise/<enterprise-id>/course/<course-key>/enroll/?catalog=<catalog-id>&utm_medium=enterprise&utm_source=test-enterprise`. This should be obtainable from the enterprise API `{{base_url}}/enterprise/api/v1/enterprise_catalogs/{{catalog_id}}/`
3. On the page, try to enroll for the course - this should do the necessary redirects, but show the an error message in the course page. Something like "You need to be enrolled to see this course content".
4. Checkout the PR branch
5. Install the PR branch `make lms-shell && pip install -e /edx/src/edx-enterprise`
6. Apply the migrations `python manage.py lms migrate`
7. Go to the [Enterprise Customer admin page](http://localhost:18000/admin/enterprise/enterprisecustomer/)
8. Check the checkbox for **Allow Enrollment in Invite Only Courses**
9. Now visit the enrollment URL as a different user and perform enrollment. The enrollment should happen and the learner should have access to the course content.

**Merge checklist:**
    - `base.in` if needed in production but edx-platform doesn't install it
    - `test-master.in` if edx-platform pins it, with a matching version
    - `make upgrade && make requirements` have been run to regenerate requirements
    - Checkout the [Database Migration](https://openedx.atlassian.net/wiki/spaces/AC/pages/23003228/Everything+About+Database+Migrations) Confluence page for helpful tips on creating migrations.
    - *Note*: This **must** be run if you modified any models.
      - It may or may not make a migration depending on exactly what you modified, but it should still be run.
    - This should be run from either a venv with all the lms/edx-enterprise requirements installed or if you checked out edx-enterprise into the src directory used by lms, you can run this command through an lms shell.
        - It would be `./manage.py lms makemigrations` in the shell.

**Post merge:**
    - *Note*: Assets will be added automatically. You just need to provide a tag (should match your version number) and title and description.
    - Each step in the release build has a condition flag that checks if the rest of the steps are done and if so will deploy to PyPi.
    (so basically once your build finishes, after maybe a minute you should see the new version in PyPi automatically (on refresh))
    - This **must** be done after the version is visible in PyPi as `make upgrade` in edx-platform will look for the latest version in PyPi.
    - Note: the edx-enterprise constraint in edx-platform **must** also be bumped to the latest version in PyPi.

[Internal Ref](https://tasks.opencraft.com/browse/BB-7619)
